File: chapters/07_compaction_and_pruning/v2_summary_compaction.py
from __future__ import annotations


def compact_with_summary(messages: list[str], keep_last: int = 3) -> dict[str, object]:
    if len(messages) <= keep_last:
        return {"summary": "", "working": list(messages)}

    old = messages[:-keep_last]
    summary = f"summary:{'; '.join(old[:3])}"
    return {
        "summary": summary,
        "working": messages[-keep_last:],
    }


def build_prompt(compacted: dict[str, object]) -> str:
    summary = str(compacted.get("summary", ""))
    working = compacted.get("working", [])
    if not isinstance(working, list):
        working = []
    return f"{summary}\nworking:{' | '.join(str(x) for x in working)}"


def main() -> None:
    msgs = [f"turn{i}" for i in range(1, 7)]
    compacted = compact_with_summary(msgs, keep_last=2)
    print(build_prompt(compacted))


if __name__ == "__main__":
    main()

File: chapters/07_compaction_and_pruning/v2_summary_compaction.ts
// TypeScript mirror for Chapter 07: Compaction and Pruning (v2)
// Run with: npx ts-node v2_summary_compaction.ts

interface Compacted {
  summary: string;
  working: string[];
}

function compactWithSummary(messages: string[], keepLast: number = 3): Compacted {
  if (messages.length <= keepLast) {
    return { summary: "", working: [...messages] };
  }

  const old = messages.slice(0, -keepLast);
  const summary = `summary:${old.slice(0, 3).join("; ")}`;
  return {
    summary,
    working: messages.slice(-keepLast),
  };
}

function buildPrompt(compacted: Compacted): string {
  const summary = compacted.summary;
  const working = compacted.working;
  return `${summary}\nworking:${working.join(" | ")}`;
}

function main(): void {
  const msgs = Array.from({ length: 6 }, (_, i) => `turn${i + 1}`);
  const compacted = compactWithSummary(msgs, 2);
  console.log(buildPrompt(compacted));
}

main();
